geometric_kernels/spaces/lie_spaces.py

```diff
@@ -0,0 +1,107 @@
+import lab as B
+import json
+from pathlib import Path
+import more_itertools
+import sympy
+from sympy.matrices.determinant import _det as sp_det
+import numpy as np
+from geometric_kernels.lab_extras import dtype_double, from_numpy
+from geometric_kernels.spaces.base import DiscreteSpectrumSpace
+from geometric_kernels.spaces.eigenfunctions import EigenfunctionWithAdditionTheorem, Eigenfunctions
+import abc
+
+
+class LieGroupAddtitionTheorem(abc.ABC, EigenfunctionWithAdditionTheorem):
+    def __init__(self, num_levels):
+        self._num_levels = num_levels
+        self._signatures, self._eigenvalues, self._degree = self.generate_signatures(self._num_levels) # both have the length num_levels
+        self._num_eigenfunctions = self.degree_to_num_eigenfunctions(self._num_levels)
+        self._characters = [LieGroupCharacter(signature) for signature in self._signatures]
+
+    def _generate_signatures(self, num_levels):
+        raise NotImplementedError
+
+    def _torus_representative(self, X):
+        """The function maps Lie Group Element X to T -- a maximal torus of the Lie group
+        [b, n, n] ---> [b, rank]"""
+        raise NotImplementedError
+
+    def _inverse(self, X):
+        """The function that computes inverse element in the group"""
+        raise NotImplementedError
+
+    def _difference(self, X: B.Numeric, X2: B.Numeric) -> B.Numeric:
+        """X -- [a,n,n], X2 -- [b,n,n] --> [a,b,n,n]"""
```

Comment:
How come `_inverse` above is abstract but multiplication here is an actual matrix product.